[CORE-12832] pp/sr: System allocator fallback for protobuf schemas

IoannisRP opened the pull request:
Implements: [CORE-12832](https://redpandadata.atlassian.net/browse/CORE-12832)

Since we don't have any control over the allocation strategy used by libproto internally, this can cause issued due to big allocations happening there.

The system allocator fallback guards have been placed where we know that big serializations are happening in libproto. Any similar call to libproto that might be introduced later on should be protected by the same guard.

Note: This PR depends on https://github.com/redpanda-data/seastar/pull/209

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none


[CORE-12832]: https://redpandadata.atlassian.net/browse/CORE-12832?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
pp/schema_registry: add fallback to system allocator for protobuf

Since we don't have any control over the allocation strategy used by
libproto internally, this can cause issued due to big allocations
happening there.

The system allocator fallback guards have been placed where we know
that big serializations are happening in libproto. Any similar
call to libproto that might be introduced later on should be protected
by the same guard.

IoannisRP committed:
pp/schema_registry: extract proto test util functions

Extract protobuf testing utility function into their own library

rockwotj commented on src/v/pandaproxy/schema_registry/test/protobuf_large_alloc.cc:
```diff
@@ -0,0 +1,109 @@
+} // namespace
+
+TEST(ProtoLargeAlloc, test_protobuf_memory_stuff) {
+    ss::memory::set_abort_on_allocation_failure(false);
```
This test will fail in sanitizer mode because the system allocator is always used. Let's do a `GTEST_SKIP()` if `SEASTAR_DEFAULT_ALLOCATOR` is defined

IoannisRP replied:
Done!

rockwotj reviewed (commented):
Looks great! Just a quick note on the test

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2a85f33e97c7bd2ceade456aa68323dc8c3b63e4..15fff932dc3837a6e6934b057218ac1c2be0e70f):
- Skip test if the system allocator is used by default.

BenPope reviewed (commented):
LGTM, didn't investigate th test failures

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/15fff932dc3837a6e6934b057218ac1c2be0e70f..c7d1677c61bba0bd2e693a02c78052e4085d485f):
- rebase to dev
- upgrade seastar version to include new system fallback allocator

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5e2d678d-e414-4cc9-9257-ac5f5404ac18-->
<!--TEST RESULTS-->
<!--82c24722-8590-4c37-b11a-06cfb9ae8fa4-->
<!--TEST RESULTS-->
<!--1d22dcc3-d1e7-40fa-9b9d-997a8d9818c5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72168">build#72168</a></summary>


| test_class         | test_method                | test_arguments         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------|:---------------------------|:-----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestOauth  | test_kafka_oauth           | {"authz_match": "acl"} | integration | https://buildkite.com/redpanda/redpanda/builds/72168#01993eb6-0a70-41b7-bc25-bd8e9c9f4cbf | FLAKY         | 20/21    | upstream reliability is '98.98305084745763'. current run reliability is '95.23809523809523'. drift is 3.74496 and the allowed drift is set to 50. The test should PASS |
| SimpleEndToEndTest | test_consumer_interruption | null                   | integration | https://buildkite.com/redpanda/redpanda/builds/72168#01993eb6-0a6c-43ca-9c27-89f05f8123b0 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| SimpleEndToEndTest | test_consumer_interruption | null                   | integration | https://buildkite.com/redpanda/redpanda/builds/72168#01993eba-9a37-450c-a712-381a44227700 | FLAKY         | 19/21    | upstream reliability is '99.27797833935018'. current run reliability is '90.47619047619048'. drift is 8.80179 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72246">build#72246</a></summary>


| test_class               | test_method                     | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:--------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_higher_level_migration_api | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72246#01994d03-33f5-4f9f-9232-84e242222e91 | FLAKY         | 20/21    | upstream reliability is '99.36708860759494'. current run reliability is '95.23809523809523'. drift is 4.12899 and the allowed drift is set to 50. The test should PASS |
| SchemaEvolutionE2ETests  | test_legal_schema_evolution     | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "add_column"}           | integration | https://buildkite.com/redpanda/redpanda/builds/72246#01994d06-60eb-4ba0-a5f8-028ec5e2a232 | FLAKY         | 20/21    | upstream reliability is '99.64285714285714'. current run reliability is '95.23809523809523'. drift is 4.40476 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations            | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72246#01994d06-60ef-4e8b-af1b-fb2e667173ce | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72327">build#72327</a></summary>


| test_class            | test_method                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite       | test_replica_pair_frequency         |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/72327#019951b2-65a6-40c5-afaa-0488710450a4 | FAIL          | 0/1      |                                                                                                                                                                        |
| PartitionBalancerTest | test_recovery_mode_rebalance_finish | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72327#019951e6-f4e3-4e20-b0a1-7b02f2c7c457 | FLAKY         | 15/21    | upstream reliability is '83.86167146974063'. current run reliability is '71.42857142857143'. drift is 12.4331 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

rockwotj commented on src/v/pandaproxy/schema_registry/test/protobuf_large_alloc.cc:
```diff
@@ -0,0 +1,113 @@
+std::vector<std::optional<memory_chunk>> fragment_memory() {
+    // Chunk size set to the same size as the maximum allocation of the chunked
+    // vector
+    constexpr auto chunk_size = 128ul * 1024ul;
```
nit: import `base/units.h` and do `128_kiB`.

rockwotj commented on src/v/pandaproxy/schema_registry/test/protobuf_large_alloc.cc:
```diff
@@ -0,0 +1,113 @@
+    GTEST_SKIP() << "Default seastar allocator detected. Skipping large alloc "
+                    "protobuf test";
+#endif
+    ss::memory::set_abort_on_allocation_failure(false);
```
let's use the scoped version or `ss::defer` to set it back, otherwise this state persists across tests (I get that there is only one test in this binary, but it's good to future proof)

rockwotj reviewed (commented):
LGTM, just a couple nits

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c7d1677c61bba0bd2e693a02c78052e4085d485f..c4940fe7ce704a2c1a3e2e7297d6668e773911b9):
- addressed nits

BenPope commented on src/v/pandaproxy/schema_registry/test/protobuf_large_alloc.cc:
```diff
@@ -0,0 +1,117 @@
+    // allocations needed for this operation
+    const auto make_proto = [&s, &sub, &large_schema]() {
+        const ss::memory::statistics start_stats = ss::memory::stats();
+        auto canocal_schema
```
I'd probably add a call to `DebugString()`, too, to ensure it succeeds under fragmentation.

IoannisRP replied:
while we don't explicitly call it in the test, it is part of the operations needed to go from text-schema to `canonical_schema`.

Are you thinking in terms of future proofing to make sure it is tested even if it is potentially removed from the canonical path?

BenPope replied:
Ahh, right. No, that's fine.

> while we don't explicitly call it in the test, it is part of the operations needed to go from text-schema to `canonical_schema`.

Ahh, of course.

> Are you thinking in terms of future proofing to make sure it is tested even if it is potentially removed from the canonical path?

No, that's fine.

BenPope commented on src/v/pandaproxy/schema_registry/test/protobuf_large_alloc.cc:
```diff
@@ -0,0 +1,117 @@
+    // allocations, as this is an implementation detail of how things are being
+    // handled in the libproto. Here we only verify that there have been some
+    // fallback allocations.
+    EXPECT_GT(make_proto(), 0);
```
We don't strictly require a fallback alloc, although we do expect one. It's a bit of an indirect way of ensuring that it's wired up to the coped allocator and that the metrics for that are working, which is fine.

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c4940fe7ce704a2c1a3e2e7297d6668e773911b9..37a6a2b0fba220d92a677c279d27d150982b1ad4):
- rebase to dev
